# Mapping Player IDs
The aim of this notebook is to create a one to one mapping between the player IDs from Kaggle Data and CMU data.
This is done by matching the first name, last name and the birth year of each player.

In [2]:
import pandas as pd
import numpy as np

# Two player sources in Kaggle Data
kagglePlayerData1 = pd.read_csv("Kaggle Data/player.csv")
kagglePlayerData2 = pd.read_csv("Kaggle Data/common_player_info.csv")
# Players from CMU Data
cmuPlayerData = pd.read_csv("Data/players.txt")

In [3]:
# convert kaggle player data's firstname lastname to lowercase
kagglePlayerData1['first_name'] = kagglePlayerData1['first_name'].str.strip().str.lower()
kagglePlayerData1['last_name'] = kagglePlayerData1['last_name'].str.strip().str.lower()
kagglePlayerData1["id"].nunique()

4831

In [4]:
kagglePlayerData2["birth_year"] = kagglePlayerData2.apply(lambda x: x["birthdate"][0:4], axis=1)
kagglePlayerData2['first_name'] = kagglePlayerData2['first_name'].str.strip().str.lower()
kagglePlayerData2['last_name'] = kagglePlayerData2['last_name'].str.strip().str.lower()
kagglePlayerData2['person_id'].nunique()

4171

In [5]:
# convert CMU player data's firstname lastname to lowercase
cmuPlayerData['firstname'] = cmuPlayerData['firstname'].str.lower().str.strip()
cmuPlayerData['lastname'] = cmuPlayerData['lastname'].str.lower().str.strip()
cmuPlayerData = cmuPlayerData.rename(columns={'firstname': 'first_name'})
cmuPlayerData = cmuPlayerData.rename(columns={'lastname': 'last_name'})
cmuPlayerData["birth_year"] = cmuPlayerData.apply(lambda x: str(x["birthdate"])[0:4], axis=1)
cmuPlayerData['ilkid'].nunique()

3572

In [6]:
# merge both the dataframes
# mergedPlayers = pd.merge(cmuPlayerData, kagglePlayerData1, left_on=['firstname', 'lastname'], right_on=['first_name','last_name'])
'''
    Using the 1st datasource from Kaggle we were not able to create a one to one mapping since it
    did not contain birth year and there were many players having the same first and last name.
'''

mergedPlayers = cmuPlayerData.merge(kagglePlayerData2, on=['first_name', 'last_name','birth_year'])
print("Total rows = "+str(len(mergedPlayers.index)))
print("Unique ilkid = "+ str(mergedPlayers["ilkid"].nunique()))
print("Unique id = "+ str(mergedPlayers["person_id"].nunique()))

mergedPlayers

Total rows = 2717
Unique ilkid = 2717
Unique id = 2717


ilkid first_name     last_name position_x  firstseason  lastseason  \
0     ABDELAL01       alaa     abdelnaby          F         1990        1994   
1     ABDULKA01     kareem  abdul-jabbar          C         1969        1988   
2     ABDULTA01      tariq   abdul-wahad          G         1997        2002   
3     ABDURSH01    shareef   abdur-rahim          F         1996        2004   
4     ABERNTO01        tom     abernethy          F         1976        1980   
...         ...        ...           ...        ...          ...         ...   
2712   ZENOTO01       tony          zeno          F         1979        1979   
2713  ZEVENPH01       phil   zevenbergen          F         1987        1987   
2714  ZIDEKGE01     george         zidek          C         1995        1997   
2715   ZOETJI01        jim          zoet          C         1982        1982   
2716  ZUNICMA01       matt         zunic          G         1948        1948   

      h_feet  h_inches  weight_x            college  ...  \
0        6.0      10.0     240.0               Duke  ...   
1        7.0       2.0     225.0               UCLA  ...   
2        6.0       6.0     223.0           Michigan  ...   
3        6.0       9.0     225.0         California  ...   
4        6.0       7.0     220.0            Indiana  ...   
...      ...       ...       ...                ...  ...   
2712     6.0       8.0     210.0      Arizona State  ...   
2713     6.0      10.0     230.0         Washington  ...   
2714     7.0       0.0     250.0               UCLA  ...   
2715     7.0       1.0     240.0         Kent State  ...   
2716     6.0       3.0     195.0  George Washington  ...   

                       playercode from_year  to_year dleague_flag nba_flag  \
0          HISTADD_alaa_abdelnaby    1990.0   1994.0            N        Y   
1     HISTADD_kareem_abdul-jabbar    1969.0   1988.0            N        Y   
2               tariq_abdul-wahad    1997.0   2003.0            N        Y   
3             shareef_abdur-rahim    1996.0   2007.0            N        Y   
4           HISTADD_tom_abernethy    1976.0   1980.0            N        Y   
...                           ...       ...      ...          ...      ...   
2712            HISTADD_tony_zeno    1979.0   1979.0            N        Y   
2713     HISTADD_phil_zevenbergen    1987.0   1987.0            N        Y   
2714                 george_zidek    1995.0   1997.0            N        Y   
2715             HISTADD_jim_zoet    1982.0   1982.0            N        Y   
2716           HISTADD_matt_zunic    1948.0   1948.0            N        Y   

     games_played_flag draft_year draft_round draft_number greatest_75_flag  
0                    Y       1990           1           25                N  
1                    Y       1969           1            1                Y  
2                    Y       1997           1           11                N  
3                    Y       1996           1            3                N  
4                    Y       1976           3           43                N  
...                ...        ...         ...          ...              ...  
2712                 Y       1979           2           32                N  
2713                 Y       1987           3           50                N  
2714                 Y       1995           1           22                N  
2715                 Y  Undrafted   Undrafted    Undrafted                N  
2716                 Y       1947         NaN          NaN                N  

[2717 rows x 43 columns]

Out of 3572 players from the CMU data, we were only able to map 2717 players with the Kaggle data. We speculate that either the names or the birth year had some discrepancy or the player just did not exit in the Kaggle data.

In [7]:
# Find out which players did not merge
merged_df = pd.merge(cmuPlayerData, mergedPlayers, on='ilkid', how='left', indicator=True)
rows_only_in_playerOfDf = merged_df[merged_df['_merge'] == 'left_only']
rows_only_in_playerOfDf


ilkid first_name_x last_name_x position  firstseason_x  \
2     ABDULMA01        mahmo  abdul-rauf        G           1990   
13    ADAMSGE01       george       adams        F           1972   
18    AINGEDA01        danny       ainge        G           1981   
19    AITCHMA01         matt       aitch        F           1967   
23    ALEKSCH01        chuck   aleksinas        C           1984   
...         ...          ...         ...      ...            ...   
3549  WYDNEAJ01         a.j.      wydner        G           1990   
3555  YONAKRI01         rich     yonakor        F           1981   
3560  YOUNGTI01          tim       young        F           1999   
3563  ZELLEDA01         dave      zeller        G           1961   
3570   ZOPFBI01         bill        zopf        G           1970   

      lastseason_x  h_feet_x  h_inches_x  weight       college_x  ...  \
2             2000       6.0         1.0   162.0             LSU  ...   
13            1974       6.0         5.0   210.0    Gardner-Webb  ...   
18            1994       6.0         4.0   175.0   Brigham Young  ...   
19            1967       6.0         7.0   230.0  Michigan State  ...   
23            1984       6.0        11.0   260.0     Connecticut  ...   
...            ...       ...         ...     ...             ...  ...   
3549          1990       6.0         2.0   180.0       Fairfield  ...   
3555          1981       6.0         9.0   220.0  North Carolina  ...   
3560          1999       7.0         0.0   220.0        Stanford  ...   
3563          1961       6.0         1.0   175.0      Miami (OH)  ...   
3570          1970       6.0         1.0   170.0        Duquesne  ...   

     from_year to_year dleague_flag nba_flag games_played_flag  draft_year  \
2          NaN     NaN          NaN      NaN               NaN         NaN   
13         NaN     NaN          NaN      NaN               NaN         NaN   
18         NaN     NaN          NaN      NaN               NaN         NaN   
19         NaN     NaN          NaN      NaN               NaN         NaN   
23         NaN     NaN          NaN      NaN               NaN         NaN   
...        ...     ...          ...      ...               ...         ...   
3549       NaN     NaN          NaN      NaN               NaN         NaN   
3555       NaN     NaN          NaN      NaN               NaN         NaN   
3560       NaN     NaN          NaN      NaN               NaN         NaN   
3563       NaN     NaN          NaN      NaN               NaN         NaN   
3570       NaN     NaN          NaN      NaN               NaN         NaN   

      draft_round  draft_number  greatest_75_flag     _merge  
2             NaN           NaN               NaN  left_only  
13            NaN           NaN               NaN  left_only  
18            NaN           NaN               NaN  left_only  
19            NaN           NaN               NaN  left_only  
23            NaN           NaN               NaN  left_only  
...           ...           ...               ...        ...  
3549          NaN           NaN               NaN  left_only  
3555          NaN           NaN               NaN  left_only  
3560          NaN           NaN               NaN  left_only  
3563          NaN           NaN               NaN  left_only  
3570          NaN           NaN               NaN  left_only  

[855 rows x 55 columns]

In [8]:
mergedPlayers.to_csv("merged_players.csv")